File: p35/SearchInsert.java
package leetcode.p35;

public class SearchInsert {
    public static int searchInsert(int[] nums, int target) {
        if (nums.length == 0) {
            return 0;
        }
        return helper(nums, 0, nums.length, target);
    }

    public static int helper(int[] nums, int start, int end, int target) {
        if (end - start <= 3) {
            for (int i = start; i < end; i++) {
                if (nums[i] == target) {
                    return i;
                } else if (nums[i] > target) {
                    return i;
                }
            }
            return end;
        }
        int index = (end + start) / 2;
        if (nums[index] == target) {
            return index;
        } else if (nums[index] > target) {
            return helper(nums, start, index, target);
        } else {
            return helper(nums, index + 1, end, target);
        }
    }

    public static void main(String[] args) {
        System.out.println(searchInsert(new int[]{1, 2, 3, 4, 5, 6, 7, 8, 9}, 10) == 9);
        System.out.println(searchInsert(new int[]{1, 3, 5, 6}, 5) == 2);
        System.out.println(searchInsert(new int[]{1, 3, 5, 6}, 2) == 1);
        System.out.println(searchInsert(new int[]{1, 3, 5, 6}, 7) == 4);

        System.out.println(searchInsert(new int[]{1, 2, 3, 4, 5, 6, 7, 8, 9}, 0) == 0);
        System.out.println(searchInsert(new int[]{1, 2, 3, 4, 5, 6, 7, 8, 9}, 1) == 0);
    }
}

File: p0/p35/search_insert.py
"""
Given a sorted array of distinct integers and a target value, return the index if the target is found. If not, return
the index where it would be if it were inserted in order.

You must write an algorithm with O(log n) runtime complexity.



Example 1:

Input: nums = [1,3,5,6], target = 5
Output: 2
Example 2:

Input: nums = [1,3,5,6], target = 2
Output: 1
Example 3:

Input: nums = [1,3,5,6], target = 7
Output: 4


Constraints:

1 <= nums.length <= 104
-104 <= nums[i] <= 104
nums contains distinct values sorted in ascending order.
-104 <= target <= 104
"""
from typing import List


class Solution:
    """
    Jul 08, 2025 16:13
    Time Complexity: O(logn)
    Space Complexity: O(1)
    """

    def searchInsert(self, nums: List[int], target: int) -> int:
        left = -1
        right = len(nums)
        while left + 1 < right:
            mid = (left + right) // 2
            if target == nums[mid]:
                return mid
            elif target < nums[mid]:
                right = mid
            else:
                left = mid

        return left + 1


if __name__ == '__main__':
    obj = Solution()
    nums = [1, 3, 5, 6]
    target = 5
    print(obj.searchInsert(nums, target) == 2)

    nums = [1, 3, 5, 6]
    target = 2
    print(obj.searchInsert(nums, target) == 1)

    nums = [1, 3, 5, 6]
    target = 7
    print(obj.searchInsert(nums, target) == 4)
